### PR TITLE
fix(gatsby-plugin-image): fix image flickers

### LINKED ISSUE #34009: [gatsby-plugin-image] art directed images flicker when parent component rerenders
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

When a parent component containing an art directed image re-renders, the art directed image "flickers" even after it has been loaded. This behaviour does not occur with a non-art directed `GatsbyImage` component.

This behaviour exists in both development and built environments.

### Reproduction Link

https://github.com/ardiewen/gatsby-image-withartdirection-repro

### Steps to Reproduce

1. Start a new repo and install `gatsby-plugin-image` `gatsby-plugin-sharp` `gatsby-transformer-sharp` `gatsby-source-filesystem` or clone this [minimal reproduction](https://github.com/ardiewen/gatsby-image-withartdirection-repro)
2. `yarn && yarn develop` to test in development environments with no flags
3. Click the increment button on each section to see the behaviour
4. `yarn build &&  yarn start` to test in the built environment
5. Click the increment button on each section to see the behaviour
...

### Expected Result

Art directed images shouldn't "flicker" after they've been loaded, unless the viewport / media query is changing.

### Actual Result

It appears that the GatsbyImage component itself isn't re-rendering, but the picture element is constantly being unmounted / remounted when the parent component is rerendered.

### Environment

```shell
System:
    OS: Linux 5.10 Ubuntu 20.04.3 LTS (Focal Fossa)
    CPU: (16) x64 AMD Ryzen 7 5800X 8-Core Processor
    Shell: 5.8 - /usr/bin/zsh
  Binaries:
    Node: 14.17.5 - ~/.nvm/versions/node/v14.17.5/bin/node
    Yarn: 1.22.15 - /usr/bin/yarn
    npm: 6.14.14 - ~/.nvm/versions/node/v14.17.5/bin/npm
  npmPackages:
    gatsby: ^4.2.0 => 4.2.0 
    gatsby-plugin-image: ^2.2.0 => 2.2.0 
    gatsby-plugin-sharp: ^4.2.0 => 4.2.0 
    gatsby-source-filesystem: ^4.2.0 => 4.2.0 
    gatsby-transformer-sharp: ^4.2.0 => 4.2.0
```

### Config Flags

N/A

### LINKED ISSUE #34190: [gatsby-plugin-image] Images flickering on route change using GatsbyImage
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

When moving from one page to another any images that are rendered using `GatsbyImage` from `gatsby-plugin-image` will briefly flicker/blink.

I have created a minimal reproduction of this which can be found attached to this issue alongside the following screen recording.

Also checked with `gatsby` and all associated plugins using version `next` but still find the same issue.

https://user-images.githubusercontent.com/8593744/144645285-5c7ea255-69cd-423c-80de-3c304bdc17e7.mov

### Reproduction Link

https://github.com/Glazy/gatsby-image-flicker-repro

### Steps to Reproduce

1. Clone and run reproduction repository
2. Use link to toggle between `/` and `/about`

### Expected Result

The image shouldn't flicker and should remain rendered at all times (like the regular `img` tag included in the example).

### Actual Result

The image flickers causing a jarring UX.

### Environment

```shell
System:
    OS: macOS 12.0
    CPU: (10) arm64 Apple M1 Pro
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 17.0.1 - ~/.nvm/versions/node/v17.0.1/bin/node
    npm: 8.1.0 - ~/.nvm/versions/node/v17.0.1/bin/npm
  Languages:
    Python: 2.7.16 - /usr/bin/python
  Browsers:
    Firefox: 94.0.2
    Safari: 15.0
  npmPackages:
    gatsby: next => 4.3.0 
    gatsby-plugin-image: next => 2.3.0 
    gatsby-plugin-sharp: next => 4.3.0 
    gatsby-source-filesystem: next => 4.3.0 
    gatsby-transformer-sharp: next => 4.3.0
```

### Config Flags

_No response_

### PR DESCRIPTION
**Pull request by wardpeet**
## Description

Fixes gatsby-plugin-image flickering between page navigations and/or state updates.

- [x] add prod

## Related Issues

Fixes #34190
Fixes #34009

### CODE REVIEW

**Comment by tyhopp** [comment]
@wardpeet I'm having a hard time separating out the parts that are fix-related from the parts that are refactored, can you add some comments (the GitHub kind, not inline 😄) that point that out? Looks great at first glance

**Comment by wardpeet** [comment]
That's the tricky bit. With the current component, we were constantly unmounting and remounting. It led to the flicker when you were doing page navigations.

The old technique is a way to do partial hydration, so we could stop hydrating gatsby-plugin-image, rely on a snippet in HTML, and only hydrate on state changes or navigations.

File: packages/gatsby-plugin-image/src/components/hooks.ts [hunk 1]
```diff
@@ -234,43 +201,14 @@ export function getMainProps(
...
-    }
-  }
-
-  // Polyfill "object-fit" if unsupported (mostly IE)
-  if (ref?.current && !(`objectFit` in document.documentElement.style)) {
-    ref.current.dataset.objectFit = style.objectFit ?? `cover`
-    ref.current.dataset.objectPosition = `${style.objectPosition ?? `50% 50%`}`
-    applyPolyfill(ref)
-  }
+  // // Polyfill "object-fit" if unsupported (mostly IE)
```
Comment (LekoArts) [inline packages/gatsby-plugin-image/src/components/hooks.ts:206]: You moved this to swapPlaceholderImage, right? So could be removed

File: packages/gatsby-plugin-image/src/components/layout-wrapper.tsx [hunk 2]
```diff
@@ -50,15 +47,17 @@ export function getSizer(
   width: number,
   height: number
 ): string {
-  let sizer: string | null = null
+  let sizer = ``
```
Comment (LekoArts) [inline packages/gatsby-plugin-image/src/components/layout-wrapper.tsx:50]: So this was the hydration bugfix, right? Did you add tests to make sure we don't regress?

Comment (wardpeet) [reply packages/gatsby-plugin-image/src/components/layout-wrapper.tsx:50]: I undid the tests that were skipped, I still need to update the test suite to actually test for flicker behavior

**Review by LekoArts — APPROVED** [review]
👍

**Pull request merged**